ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -316,229 +74,163 @@ private static PhysicalSchema getPhysicalSchema(final SourceStep<?> streamSource
+        return transformed.mapValues(row -> row, materialized);
+
       } else {
         // if we know this table source is repartitioned later in the topology,
         // we do not need to force a materialization at this source step since the
         // re-partitioned topic will be used for any subsequent state stores, in lieu
         // of the original source topic, thus avoiding the issues above.
         // See https://github.com/confluentinc/ksql/issues/6650
-        table = source.mapValues(row -> row);
+        return transformed.mapValues(row -> row);
       }
-    }
 
-    return table
-        .transformValues(new AddKeyAndTimestampColumns<>(keyGenerator));
   }
 
-  private static StaticTopicSerde.Callback getRegisterCallback(
+  @Override
+  <K> KTable<K, GenericRow> buildWindowedKTable(SourceStep<?> streamSource,
       final RuntimeBuildContext buildContext,
-      final FormatInfo valueFormat
+      final Consumed<K, GenericRow> consumed,
+      final Function<K, Collection<?>> keyGenerator,
+      final Materialized<K, GenericRow, KeyValueStore<Bytes, byte[]>> materialized,
+      final Serde<GenericRow> valueSerde,
+      final String stateStoreName,
+      final PlanInfo planInfo
   ) {
-    final boolean schemaRegistryEnabled = !buildContext
-        .getKsqlConfig()
-        .getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY)
-        .isEmpty();
-
-    final boolean useSR = FormatFactory
-        .fromName(valueFormat.getFormat())
-        .supportsFeature(SerdeFeature.SCHEMA_INFERENCE);
-
-    if (!schemaRegistryEnabled || !useSR) {
-      return (t1, t2, data) -> { };
-    }
+    final String changelogTopic = changelogTopic(buildContext, stateStoreName);
+    final Callback onFailure = getRegisterCallback(
+        buildContext, streamSource.getFormats().getValueFormat());
+
+    final KTable<K, GenericRow> source = buildContext
+        .getStreamsBuilder()
+        .table(
+            streamSource.getTopicName(),
+            consumed.withValueSerde(StaticTopicSerde.wrap(changelogTopic, valueSerde, onFailure))
+        );
 
-    return new RegisterSchemaCallback(buildContext.getServiceContext().getSchemaRegistryClient());
+    return source.transformValues(new AddKeyAndPseudoColumns<>(
+        keyGenerator, streamSource.getPseudoColumnVersion()), materialized);
   }
 
-  /**
-   * This code mirrors the logic that generates the name for changelog topics
-   * in kafka streams, which follows the pattern:
-   * <pre>
-   *    applicationID + "-" + stateStoreName + "-changelog".
-   * </pre>
-   */
-  private static String changelogTopic(
+  @Override
+  Materialized<GenericKey, GenericRow, KeyValueStore<Bytes, byte[]>> buildTableMaterialized(
+      final SourceStep<KTableHolder<GenericKey>> source,
       final RuntimeBuildContext buildContext,
-      final String stateStoreName
+      final MaterializedFactory materializedFactory,
+      final Serde<GenericKey> keySerde,
+      final Serde<GenericRow> valueSerde
```

Comment:
Clarify that these are the key and value serde for the source, which is why they should not be used for the Materialized:
```suggestion
      final Serde<GenericKey> sourceKeySerde,
      final Serde<GenericRow> sourceValueSerde
```